### PR TITLE
Add new lambda runtimes and more parity fixes

### PR DESCRIPTION
**Pull request by joe4dev**
Superseeds https://github.com/localstack/localstack/pull/9659 (Node 20) and https://github.com/localstack/localstack/pull/9685 (Java 20)
Big kudos to our community members @ataylorme and @eddumelendez for your initiative and efforts 👏👏👏 🥳 

## Motivation

AWS recently announced new Lambda runtimes (based on [AL2023](https://aws.amazon.com/blogs/compute/introducing-the-amazon-linux-2023-runtime-for-aws-lambda/) such as [Nodejs 20.x](https://aws.amazon.com/blogs/compute/node-js-20-x-runtime-now-available-in-aws-lambda/)) and our currently supported runtimes have become outdated. Furthermore, more runtime changes including deprecations and new runtimes are expected to come in the future.

Updating runtimes is currently too cumbersome and many effects (both from us and contributors) are incomplete and need to be followed up by bugfixes.

## Changes

This PR fixes AWS-parity for Lambda and updates many snapshots:

* Add the new Lambda runtimes `nodejs20.x`, `python3.12` (not yet in the docs but already live), `java21`, 
`provided.al2023` (see https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html)
* Centralize everything related to Lambda runtimes to simplify future updates
* Add the deprecated runtimes `nodejs12.x` and `dotnetcore3.1` to a `DEPRECATED_RUNTIMES` list but keep them available in LocalStack for now. We log a deprecation warning upon creating or updating deprecated runtimes.
* Add a new marker `lambda_runtime_update` to filter test cases related to Lambda runtime updates for easier snapshot updates. This allows us to keep the tests grouped logically and easily update snapshots related to runtime updates.
* Update many Lambda snapshots
* Fix new AWS behavior enforcing stricter checks for large ZIP files (see `RequestEntityTooLargeException`) and a new higher limit for request size. Like in AWS, users will see the clearer zipped file size exception rather than the more confusing request size too large exception.
* Fix updated validations
* Add skip_snapshot_verify for some new AWS features (e.g. InvokeMode)
* Fix timeout against AWS for `test_vpc_config`

## Idea for v4

Having an opt-in flag such as `LAMBDA_EXTENDED_RUNTIMES` would enable deprecated as well as upcoming preview runtimes while maintaining full AWS parity.

## TODO

What's left to do:

- [x] Follow up failing tests after full test run
- [x] Greenify ext PR

## Follow Ups

- [ ]  Suggest new extended runtime support feature ⇒ such an opt-in is a major change planned for v4. Kind of `LAMBDA_EXTENDED_RUNTIMES`
- [ ]  Update [config docs](https://docs.localstack.cloud/references/configuration/#lambda) for `LAMBDA_LIMITS_CREATE_FUNCTION_REQUEST_SIZE` to new default `70_167_211`

### CODE REVIEW

File: localstack/config.py [hunk 1]
```diff
@@ -904,7 +904,7 @@ def populate_edge_configuration(
 )
 # PUBLIC: documented upon customer request
 LAMBDA_LIMITS_CREATE_FUNCTION_REQUEST_SIZE = int(
-    os.environ.get("LAMBDA_LIMITS_CREATE_FUNCTION_REQUEST_SIZE", 69_905_067)
+    os.environ.get("LAMBDA_LIMITS_CREATE_FUNCTION_REQUEST_SIZE", 70_167_211)
```
Comment (dfangl) [inline localstack/config.py:907]: Why did this change, and where does the new value come from?

File: localstack/services/lambda_/provider.py [hunk 2]
```diff
@@ -1795,6 +1821,13 @@ def list_event_source_mappings(
     # ============ FUNCTION URLS ============
     # =======================================
 
+    @staticmethod
+    def _validate_qualifier(qualifier: str) -> None:
+        if qualifier == "$LATEST" or (qualifier and api_utils.qualifier_is_version(qualifier)):
+            raise ValidationException(
+                f"1 validation error detected: Value '{qualifier}' at 'qualifier' failed to satisfy constraint: Member must satisfy regular expression pattern: ((?!^\\d+$)^[0-9a-zA-Z-_]+$)"
+            )
+
```
Comment (dfangl) [inline localstack/services/lambda_/provider.py:1830]: Nice!

File: tests/aws/services/lambda_/dynamodb-local-metadata.json
Comment (dfangl) [inline tests/aws/services/lambda_/dynamodb-local-metadata.json:1]: I think this file should not have been committed.

**Review by dfangl — APPROVED** [review]
LGTM, great set of changes! I think the marker for tests that need to be run for an update is a great idea!

File: tests/aws/services/lambda_/dynamodb-local-metadata.json
Comment (joe4dev) [reply tests/aws/services/lambda_/dynamodb-local-metadata.json:1]: ups, fixed.
Thanks

File: localstack/config.py (hunk 1 above)
Comment (joe4dev) [reply localstack/config.py:907]: It comes from the updated snapshots.

AWS changes the behavior related to these limits. Therefore, I added an extra test and adjusted the tests to cover special cases (e.g., zip within limit but request size exceeding).

My guess: The previous error message was too confusing. People uploaded large zips and got a message about request size:
`Request must be smaller than 70167211 bytes for the CreateFunction operation`

Now they check the zip (if relevant) before checking the request size limit leading to a clearer message:
`Zipped size must be smaller than 52428800 bytes`

File: tests/aws/services/lambda_/test_lambda_api.snapshot.json [hunk 3]
```diff
@@ -5653,12 +5655,12 @@
...
   },
   "tests/aws/services/lambda_/test_lambda_api.py::TestLambdaTags::test_tag_exceptions": {
-    "recorded-date": "30-09-2022, 15:36:19",
+    "recorded-date": "20-11-2023, 17:02:01",
     "recorded-content": {
       "tag_lambda_invalidarn": {
         "Error": {
           "Code": "ValidationException",
-          "Message": "1 validation error detected: Value 'arn:aws:something' at 'resource' failed to satisfy constraint: Member must satisfy regular expression pattern: arn:(aws[a-zA-Z-]*)?:lambda:[a-z]{2}((-gov)|(-iso(b?)))?-[a-z]+-\\d{1}:\\d{12}:function:[a-zA-Z0-9-_]+(:(\\$LATEST|[a-zA-Z0-9-_]+))?"
+          "Message": "1 validation error detected: Value 'arn:aws:something' at 'resource' failed to satisfy constraint: Member must satisfy regular expression pattern: arn:(aws[a-zA-Z-]*)?:lambda:[a-z]{2}((-gov)|(-iso(b?)))?-[a-z]+-\\d{1}:\\d{12}:(function:[a-zA-Z0-9-_]+(:(\\$LATEST|[a-zA-Z0-9-_]+))?|layer:[a-zA-Z0-9-_]+)"
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.snapshot.json:5663]: I didn't look into implications for the implementation here

File: localstack/config.py (hunk 1 above)
Comment (dfangl) [reply localstack/config.py:907]: Ah, great! Thanks for clarifying.

**Pull request merged**